File: telemetry/aitf/sdk/python/src/aitf/processors/pii_processor.py
"""AITF PII Detection Processor.

OTel SpanProcessor that detects and optionally redacts Personally
Identifiable Information (PII) in AI inputs and outputs.

Based on PII detection patterns from AITelemetry project.
"""

from __future__ import annotations

import hashlib
import hmac
import os
import re
from dataclasses import dataclass
from typing import Any

from opentelemetry.context import Context
from opentelemetry.sdk.trace import ReadableSpan, Span, SpanProcessor

from aitf.semantic_conventions.attributes import SecurityAttributes

# Maximum content length to analyze (prevent unbounded CPU usage)
_MAX_CONTENT_LENGTH = 100_000

# PII detection patterns
# Note: Credit card pattern uses explicit groups instead of nested
# quantifiers to prevent ReDoS
PII_PATTERNS: dict[str, re.Pattern] = {
    "email": re.compile(
        r"\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b"
    ),
    "phone": re.compile(
        r"\b(?:\+?1[-.\s]?)?\(?\d{3}\)?[-.\s]?\d{3}[-.\s]?\d{4}\b"
    ),
    "ssn": re.compile(
        r"\b\d{3}-\d{2}-\d{4}\b"
    ),
    "credit_card": re.compile(
        # Explicit repetition instead of nested quantifier {3} to prevent ReDoS
        r"\b\d{4}[-\s]?\d{4}[-\s]?\d{4}[-\s]?\d{4}\b"
    ),
    "api_key": re.compile(
        r"\b(?:sk-|pk-|ak-|key-)[A-Za-z0-9]{20,}\b"
    ),
    "jwt": re.compile(
        r"\beyJ[A-Za-z0-9_-]+\.eyJ[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\b"
    ),
    "ip_address": re.compile(
        r"\b(?:(?:25[0-5]|2[0-4]\d|[01]?\d\d?)\.){3}(?:25[0-5]|2[0-4]\d|[01]?\d\d?)\b"
    ),
    "password": re.compile(
        r"(?:password|passwd|pwd)\s*[=:]\s*\S+", re.IGNORECASE
    ),
    "aws_key": re.compile(
        r"\bAKIA[0-9A-Z]{16}\b"
    ),
}


@dataclass
class PIIDetection:
    """A PII detection result."""

    pii_type: str
    count: int
    locations: list[tuple[int, int]]  # (start, end) positions


class PIIProcessor(SpanProcessor):
    """OTel SpanProcessor that detects PII in AI span content.

    Can operate in three modes:
    - "flag": Detect PII and add span attributes (default)
    - "redact": Replace PII with [REDACTED] placeholder
    - "hash": Replace PII with HMAC-SHA256 hash (keyed, non-reversible)

    Usage:
        provider.add_span_processor(PIIProcessor(
            detect_types=["email", "ssn", "credit_card", "api_key"],
            action="redact",
        ))
    """

    def __init__(
        self,
        detect_types: list[str] | None = None,
        action: str = "flag",
        custom_patterns: dict[str, re.Pattern] | None = None,
        hash_key: bytes | None = None,
    ):
        self._detect_types = detect_types or list(PII_PATTERNS.keys())
        self._action = action
        self._patterns: dict[str, re.Pattern] = {}

        # Generate a random HMAC key for PII hashing if not provided
        # This ensures hashes are unique per processor instance and
        # cannot be reversed by an attacker who knows the algorithm
        self._hash_key = hash_key or os.urandom(32)

        # Build active pattern set
        for pii_type in self._detect_types:
            if pii_type in PII_PATTERNS:
                self._patterns[pii_type] = PII_PATTERNS[pii_type]

        # Validate and add custom patterns
        if custom_patterns:
            for name, pattern in custom_patterns.items():
                if not isinstance(pattern, re.Pattern):
                    raise TypeError(
                        f"Custom pattern '{name}' must be a compiled re.Pattern"
                    )
                # Verify the pattern compiles and test it with empty string
                # to detect obvious issues
                try:
                    pattern.search("")
                except re.error as e:
                    raise ValueError(
                        f"Custom pattern '{name}' is invalid: {e}"
                    ) from e
                self._patterns[name] = pattern

    def on_start(self, span: Span, parent_context: Context | None = None) -> None:
        pass

    def on_end(self, span: ReadableSpan) -> None:
        """Detect PII in span content."""
        attrs = span.attributes or {}
        if not any(key.startswith(("gen_ai.", "agent.", "mcp.", "skill.", "security.", "cost.")) for key in attrs.keys()):
            return

        all_detections: list[PIIDetection] = []

        # Check span events for content
        for event in span.events or []:
            event_attrs = event.attributes or {}
            for val in event_attrs.values():
                if isinstance(val, str):
                    detections = self.detect_pii(val)
                    all_detections.extend(detections)

        # Check relevant span attributes
        for key in (
            "gen_ai.tool.call.arguments",
            "gen_ai.tool.call.result",
            "skill.input",
            "skill.output",
        ):
            val = attrs.get(key)
            if isinstance(val, str):
                detections = self.detect_pii(val)
                all_detections.extend(detections)

    def detect_pii(self, text: str) -> list[PIIDetection]:
        """Detect PII in text. Returns list of PIIDetection objects."""
        # Truncate to prevent excessive CPU usage
        if len(text) > _MAX_CONTENT_LENGTH:
            text = text[:_MAX_CONTENT_LENGTH]

        detections: list[PIIDetection] = []
        for pii_type, pattern in self._patterns.items():
            matches = list(pattern.finditer(text))
            if matches:
                detections.append(PIIDetection(
                    pii_type=pii_type,
                    count=len(matches),
                    locations=[(m.start(), m.end()) for m in matches],
                ))
        return detections

    def redact_pii(self, text: str) -> tuple[str, list[PIIDetection]]:
        """Detect and redact PII from text.

        Returns (redacted_text, detections).
        """
        detections = self.detect_pii(text)
        if not detections:
            return text, []

        result = text
        for detection in sorted(detections, key=lambda d: d.locations[0][0], reverse=True):
            for start, end in sorted(detection.locations, reverse=True):
                original = result[start:end]
                if self._action == "redact":
                    replacement = f"[{detection.pii_type.upper()}_REDACTED]"
                elif self._action == "hash":
                    # Use HMAC-SHA256 with instance-specific key for secure hashing
                    hash_val = hmac.new(
                        self._hash_key,
                        original.encode(),
                        hashlib.sha256,
                    ).hexdigest()[:16]
                    replacement = f"[{detection.pii_type.upper()}:{hash_val}]"
                else:
                    continue
                result = result[:start] + replacement + result[end:]

        return result, detections

    def get_pii_summary(self, text: str) -> dict[str, Any]:
        """Get a summary of PII detected in text."""
        detections = self.detect_pii(text)
        if not detections:
            return {
                SecurityAttributes.PII_DETECTED: False,
                SecurityAttributes.PII_TYPES: [],
                SecurityAttributes.PII_COUNT: 0,
                SecurityAttributes.PII_ACTION: self._action,
            }
        return {
            SecurityAttributes.PII_DETECTED: True,
            SecurityAttributes.PII_TYPES: [d.pii_type for d in detections],
            SecurityAttributes.PII_COUNT: sum(d.count for d in detections),
            SecurityAttributes.PII_ACTION: self._action,
        }

    def shutdown(self) -> None:
        pass

    def force_flush(self, timeout_millis: int = 30000) -> bool:
        return True

File: telemetry/aitf/sdk/typescript/src/processors/pii-processor.ts
/**
 * AITF PII Detection Processor.
 *
 * OTel SpanProcessor that detects and optionally redacts Personally
 * Identifiable Information (PII) in AI inputs and outputs.
 *
 * Based on PII detection patterns from AITelemetry project.
 */

import { createHmac, randomBytes } from "crypto";
import { Context, Span } from "@opentelemetry/api";
import { ReadableSpan, SpanProcessor } from "@opentelemetry/sdk-trace-base";
import { SecurityAttributes } from "../semantic-conventions/attributes";

/** Maximum content length to analyze (prevent unbounded CPU usage). */
const MAX_CONTENT_LENGTH = 100_000;

/** PII detection patterns for 9 PII types. */
// Note: Credit card pattern uses explicit groups instead of nested
// quantifiers to prevent ReDoS
const PII_PATTERNS: Record<string, RegExp> = {
  email: /\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}\b/g,
  phone: /\b(?:\+?1[-.\s]?)?\(?\d{3}\)?[-.\s]?\d{3}[-.\s]?\d{4}\b/g,
  ssn: /\b\d{3}-\d{2}-\d{4}\b/g,
  // Explicit repetition instead of nested quantifier {3} to prevent ReDoS
  credit_card: /\b\d{4}[-\s]?\d{4}[-\s]?\d{4}[-\s]?\d{4}\b/g,
  api_key: /\b(?:sk-|pk-|ak-|key-)[A-Za-z0-9]{20,}\b/g,
  jwt: /\beyJ[A-Za-z0-9_-]+\.eyJ[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\b/g,
  ip_address:
    /\b(?:(?:25[0-5]|2[0-4]\d|[01]?\d\d?)\.){3}(?:25[0-5]|2[0-4]\d|[01]?\d\d?)\b/g,
  password: /(?:password|passwd|pwd)\s*[=:]\s*\S+/gi,
  aws_key: /\bAKIA[0-9A-Z]{16}\b/g,
};

/** A PII detection result. */
export interface PIIDetection {
  piiType: string;
  count: number;
  locations: Array<[number, number]>; // [start, end] positions
}

/** Options for configuring the PIIProcessor. */
export interface PIIProcessorOptions {
  /** Which PII types to detect. Defaults to all 9 types. */
  detectTypes?: string[];
  /** Action to take: "flag", "redact", or "hash". Defaults to "flag". */
  action?: "flag" | "redact" | "hash";
  /** Custom patterns to add. Keys are validated to prevent prototype pollution. */
  customPatterns?: Record<string, RegExp>;
  /** HMAC key for PII hashing. Auto-generated if not provided. */
  hashKey?: Buffer;
}

/**
 * OTel SpanProcessor that detects PII in AI span content.
 *
 * Can operate in three modes:
 * - "flag": Detect PII and add span attributes (default)
 * - "redact": Replace PII with [REDACTED] placeholder
 * - "hash": Replace PII with HMAC-SHA256 hash (keyed, non-reversible)
 *
 * Usage:
 *   provider.addSpanProcessor(new PIIProcessor({
 *     detectTypes: ["email", "ssn", "credit_card", "api_key"],
 *     action: "redact",
 *   }));
 */
export class PIIProcessor implements SpanProcessor {
  private readonly _detectTypes: string[];
  private readonly _action: "flag" | "redact" | "hash";
  private readonly _patterns: Record<string, RegExp>;
  private readonly _hashKey: Buffer;

  constructor(options: PIIProcessorOptions = {}) {
    this._detectTypes = options.detectTypes ?? Object.keys(PII_PATTERNS);
    this._action = options.action ?? "flag";
    this._patterns = {};

    // Generate a random HMAC key for PII hashing if not provided
    this._hashKey = options.hashKey ?? randomBytes(32);

    // Build active pattern set
    for (const piiType of this._detectTypes) {
      if (Object.prototype.hasOwnProperty.call(PII_PATTERNS, piiType)) {
        this._patterns[piiType] = PII_PATTERNS[piiType];
      }
    }

    // Validate and add custom patterns (prevent prototype pollution)
    if (options.customPatterns) {
      for (const [name, pattern] of Object.entries(options.customPatterns)) {
        // Skip prototype-inherited properties
        if (!Object.prototype.hasOwnProperty.call(options.customPatterns, name)) {
          continue;
        }
        // Reject dangerous property names
        if (name === "__proto__" || name === "constructor" || name === "prototype") {
          throw new Error(`Invalid custom pattern name: '${name}'`);
        }
        if (!(pattern instanceof RegExp)) {
          throw new TypeError(`Custom pattern '${name}' must be a RegExp instance`);
        }
        this._patterns[name] = pattern;
      }
    }
  }

  onStart(_span: Span, _parentContext: Context): void {
    // No-op
  }

  onEnd(span: ReadableSpan): void {
    const attrs = span.attributes ?? {};

    const hasAiPrefix = Object.keys(attrs).some(
      (key) => key.startsWith("gen_ai.") || key.startsWith("")
    );
    if (!hasAiPrefix) {
      return;
    }

    const allDetections: PIIDetection[] = [];

    // Check span events for content
    for (const event of span.events ?? []) {
      const eventAttrs = event.attributes ?? {};
      for (const val of Object.values(eventAttrs)) {
        if (typeof val === "string") {
          const detections = this.detectPII(val);
          allDetections.push(...detections);
        }
      }
    }

    // Check relevant span attributes
    for (const key of [
      "mcp.tool.input",
      "mcp.tool.output",
      "skill.input",
      "skill.output",
    ]) {
      const val = attrs[key];
      if (typeof val === "string") {
        const detections = this.detectPII(val);
        allDetections.push(...detections);
      }
    }
  }

  /**
   * Detect PII in text. Returns list of PIIDetection objects.
   */
  detectPII(text: string): PIIDetection[] {
    // Truncate to prevent excessive CPU usage
    if (text.length > MAX_CONTENT_LENGTH) {
      text = text.slice(0, MAX_CONTENT_LENGTH);
    }

    const detections: PIIDetection[] = [];

    for (const [piiType, pattern] of Object.entries(this._patterns)) {
      // Reset regex state for global patterns
      const regex = new RegExp(pattern.source, pattern.flags);
      const matches: Array<[number, number]> = [];
      let match: RegExpExecArray | null;

      while ((match = regex.exec(text)) !== null) {
        matches.push([match.index, match.index + match[0].length]);
      }

      if (matches.length > 0) {
        detections.push({
          piiType,
          count: matches.length,
          locations: matches,
        });
      }
    }

    return detections;
  }

  /**
   * Detect and redact PII from text.
   * Returns [redacted_text, detections].
   */
  redactPII(text: string): [string, PIIDetection[]] {
    const detections = this.detectPII(text);
    if (detections.length === 0) {
      return [text, []];
    }

    let result = text;

    // Sort detections by first location, process in reverse order
    const sortedDetections = [...detections].sort(
      (a, b) => a.locations[0][0] - b.locations[0][0]
    );

    for (const detection of sortedDetections.reverse()) {
      const sortedLocations = [...detection.locations].sort(
        (a, b) => b[0] - a[0]
      );
      for (const [start, end] of sortedLocations) {
        const original = result.slice(start, end);
        let replacement: string;

        if (this._action === "redact") {
          replacement = `[${detection.piiType.toUpperCase()}_REDACTED]`;
        } else if (this._action === "hash") {
          // Use HMAC-SHA256 with instance-specific key for secure hashing
          const hashVal = createHmac("sha256", this._hashKey)
            .update(original)
            .digest("hex")
            .slice(0, 16);
          replacement = `[${detection.piiType.toUpperCase()}:${hashVal}]`;
        } else {
          continue;
        }

        result = result.slice(0, start) + replacement + result.slice(end);
      }
    }

    return [result, detections];
  }

  /**
   * Get a summary of PII detected in text.
   */
  getPIISummary(
    text: string
  ): Record<string, boolean | string[] | number | string> {
    const detections = this.detectPII(text);

    if (detections.length === 0) {
      return {
        [SecurityAttributes.PII_DETECTED]: false,
        [SecurityAttributes.PII_TYPES]: [] as string[],
        [SecurityAttributes.PII_COUNT]: 0,
        [SecurityAttributes.PII_ACTION]: this._action,
      };
    }

    return {
      [SecurityAttributes.PII_DETECTED]: true,
      [SecurityAttributes.PII_TYPES]: detections.map((d) => d.piiType),
      [SecurityAttributes.PII_COUNT]: detections.reduce(
        (sum, d) => sum + d.count,
        0
      ),
      [SecurityAttributes.PII_ACTION]: this._action,
    };
  }

  async shutdown(): Promise<void> {
    // No-op
  }

  async forceFlush(): Promise<void> {
    // No-op
  }
}
